Address translation with exception for hosts on current control connection

wprzytula opened the pull request:
To be consistent with other drivers' behaviour, as well as to mitigate the _possibly empty rpc_address_ problem, we turned off address translation for hosts that serve as metadata providers on current control connection. 
This re-reverts #471 and provides necessary fixes not to cause #487.
Fixes: #490 #487 #470 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [ ] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

wprzytula commented:
Integration tests of this feature could be written using proxy, so they are not included in this PR yet.

wprzytula commented:
@psarna @cvybhu waiting for review.

cvybhu reviewed (approved):
lgtm

psarna commented on scylla/src/transport/cluster.rs:
```diff
@@ -52,7 +52,7 @@ pub struct Datacenter {
 #[derive(Clone)]
 pub struct ClusterData {
-    pub(crate) initially_known_peers: Arc<Vec<SocketAddr>>, // Invariant: nonempty after Cluster::new()
+    pub(crate) initially_known_peers: Vec<SocketAddr>, // Invariant: nonempty after Cluster::new()
```
it looks like this commit fixes an issue introduced in another commit in the same series (c12026a3acff585f20a2c8472bd9ecf562b4d6ec) - if that's indeed the case, please squash them together

piodul reviewed (changes requested):
Needs rebase.

I'm not sure if the logic is correct. You are keeping the list of _initial_ contact points and use them to decide whether the address returned by the metadata fetch should be translated or not. However, consider this scenario:

- Create a session with initial contact points [A],
- The session establishes a control connection to the node A, learns that there are nodes [A, B, C], and the application starts doing its stuff,
- The node A goes down and the session tries to reestablish the control connection to either node B or C. This will not work well because, when connecting to node X, the metadata fetch will return IP for the node X which must not be translated. Because neither B or C is in the initially known peers list, the driver will attempt to translate their addresses and will fail to do so, which will probably result in no queries being routed to that node (issue https://github.com/scylladb/scylla-rust-driver/issues/487).

IIUC only one address returned by metadata fetch should _not_ be translated: the address of the control connection. How about not keeping the initial peers list and filtering out based on the current control connection address?

Another option is to include information whether the peer's address needs translation or not in the `Peer` structure. This information is known during metadata fetch, when the `Peer` structs are constructed. I'm not sure which one is better; on one hand, the second solution requires changing a public struct and may break more code, on the other hand this information may actually be interesting to the users.

wprzytula commented on scylla/src/transport/session.rs:
```diff
@@ -59,7 +59,7 @@ use openssl::ssl::SslContext;
 #[derive(Debug, Copy, Clone)]
 pub enum TranslationError {
-    NoRuleForAddress,
+    NoRuleForAddress(SocketAddr),
```
Okay, right. It is probably better to leave responsibility of being aware of which address is being translated to the caller.

psarna commented:
@wprzytula I see that @piodul had a long comment and you pushed some changes. Did they address @piodul's comments? Please post a short changelog of what's new in the series.

wprzytula commented:
v2:
- reworked the fix so that now only current control connection host's address is not translated,
- removed the unnecessary data from `NoRuleForAddress`.

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -42,6 +42,8 @@ pub(crate) struct MetadataReader {
 pub struct Metadata {
     pub peers: Vec<Peer>,
     pub keyspaces: HashMap<String, Keyspace>,
+    // This is not really part of metadata, but is required for their further interpretation.
+    pub control_connection_address: SocketAddr,
```
The more I think about it, the more it makes sense to me to put separate translated/untranslated addresses into the `Peer` struct. With the current implementation in the PR, interpreting the address in a Peer always requires the parent Metadata to be known (or at least the control connection address from the moment the metadata was created), which is a bit cumbersome.

Keeping both translated/untranslated addresses in the Peer would translate to an easier API. You could pass an `Option<&dyn AddressTranslator>` as an argument to `query_peers` and translate the address when creating a `Peer`. A positive side effect of that is that you will only have to translate the address once, when fetching - not twice (`ClusterData::new` + `MetadataReader::update_known_peers`), and you don't need to keep the control_connection_address used during the metadata fetch.

I suggest adding an `untranslated_address: Option<IpAddr>` to the peer. The existing `address` would represent the translated address and would work as it works now in the PR. The new `untranslated_address` would always be equal to the address in the `system.peers` for non-control-connection nodes; for the control connection node, when address translation is disabled we can use the address from `system.local`, and `None` otherwise (as the address we used to connect is most likely different and we just don't know).

Finally, there is a chicken-and-egg problem here as the translator requires a `Peer` to translate the address... Let's introduce:

```rust
#[non_exhaustive] // <- so that we can add more fields in a backwards-compatible way
pub struct UntranslatedPeer {
    pub untranslated_address: SocketAddr,
}
```

and use it as an argument instead.

wprzytula replied:
At first glance, your idea seems to work very well. I'll try to implement it.

wprzytula replied:
> I suggest adding an untranslated_address: Option<IpAddr> to the peer.
Did you mean `Option<SocketAddr>` ?

piodul replied:
> Did you mean `Option<SocketAddr>` ?

Yes, my mistake. Edited the comment.

wprzytula commented:
v3:
- AddressTranslator now accepts UntranslatedPeer
- Peer was extended with untranslated_address field
- address translation occurs exactly once per metadata fetch, in `query_peers()`
- split into commits more reasonably

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -408,55 +420,88 @@ async fn query_peers(conn: &Connection, connect_port: u16) -> Result<Vec<Peer>,
+        .chain(typed_local_rows.map(|res| res.map(|local_row| (true, local_row))));
+
+    let translated_peers_futures = untranslated_rows.map(|untranslated_row| async {
+        match untranslated_row {
```
You can use `?` operator in closures and futures which return `Result`. You can reduce nesting by rewriting the first match as:

```rust
let (is_local, (untranslated_ip_addr, datacenter, rack, tokens)) = untranslated_row.map_err(|_| {
    QueryError::ProtocolError("system.peers or system.local has invalid column type")
})?;

// Rest of the code for the `Ok(_)` case
```

The resulting code will be flatter and the diff will become smaller.

However, `rustc` will need your help in order to determine the return type of the `async` block because it won't be able to infer the `E` in `Result<T, E>`. It won't be able to do it from the `?` expression because the `?` operator is allowed to convert the error type with an `Into` conversion, and you won't be returning any `Err(x)` variants where `x` would be a known type.

I don't think it is possible to annotate the return type of an async block directly - however, you can use the turbofish operator to specify `E` when returning the `Ok` variant, e.g.:

```rust
return Result::<_, QueryError>::Ok(None);
```

wprzytula commented:
v4: reduced level of nesting in `query_peers()`
